### PR TITLE
feat: web3auth integration

### PR DESCRIPTION
**Pull request by JosesGabriel**

### CODE REVIEW

File: lib/internal/string_constants.dart
Comment (kabaluyot) [inline lib/internal/string_constants.dart:1]: move to .env

WEB3AUTH_APP_BUNDLE_ID=
WEB3AUTH_APP_URL_SCHEME=

File: lib/infrastructures/service/cubit/web3_cubit.dart [hunk 1]
```diff
@@ -12,31 +14,38 @@ class Web3Cubit extends Cubit<Web3State> {
   Web3Cubit({
     required this.web3Client,
     required this.greeterContract,
+    required this.storage,
   }) : super(const Web3State());
 
   // core declarations
   final Web3Client web3Client;
   final DeployedContract greeterContract;
```
Comment (kabaluyot) [inline lib/infrastructures/service/cubit/web3_cubit.dart:22]: group by modifiers

final ...
final ...
late ...
late ...

File: lib/infrastructures/service/cubit/web3_cubit.dart [hunk 2]
```diff
@@ -12,31 +14,38 @@ class Web3Cubit extends Cubit<Web3State> {
...
   /// Terminates metamask, provider, contract connections
-  void closeConnection() {
+  void closeConnection(LoginType loginType) {
     fetchGreetingTimer.cancel();
-    walletConnector.killSession();
-    walletConnector.close();
+    if (loginType == LoginType.metaMask) {
+      walletConnector?.killSession();
+      walletConnector?.close();
+    } else if (loginType == LoginType.google) {
```
Comment (kabaluyot) [inline lib/infrastructures/service/cubit/web3_cubit.dart]: loginType (=> provider) should be:

- metaMask
- web3Auth

google provider or other social/email providers are all under web3auth provider

File: lib/infrastructures/service/cubit/web3_cubit.dart [hunk 3]
```diff
@@ -50,10 +59,28 @@ class Web3Cubit extends Cubit<Web3State> {
     fetchGreetingTimer =
```
Comment (kabaluyot) [inline lib/infrastructures/service/cubit/web3_cubit.dart]: remove na goro nato ning fetch-interval related diri kols? Introduce nalang ta pull to refresh ..

most likely dapps via websocket or manual refresh for data

File: lib/infrastructures/service/cubit/web3_cubit.dart [hunk 4]
```diff
@@ -50,10 +59,28 @@ class Web3Cubit extends Cubit<Web3State> {
...
         Timer.periodic(const Duration(seconds: 5), (_) => fetchGreeting());
 
-    emit(InitializeProviderSuccess(
+    emit(InitializeMetaMaskProviderSuccess(
         accountAddress: sender, networkName: getNetworkName(session.chainId)));
   }
 
+  /// Initialize Google provider
+  Future<void> initializeGoogleProvider() async {
```
Comment (kabaluyot) [inline lib/infrastructures/service/cubit/web3_cubit.dart]: should be initializeWeb3AuthProvider

All social logins supported by web3auth at the end, will be invoking this

File: lib/infrastructures/service/cubit/web3_cubit.dart [hunk 5]
```diff
@@ -71,36 +98,60 @@ class Web3Cubit extends Cubit<Web3State> {
...
-          Duration(milliseconds: getBlockTime(sessionStatus.chainId)),
-          (_) async {
-        TransactionReceipt? t = await web3Client.getTransactionReceipt(txnHash);
-        if (t != null) {
-          emit(const UpdateGreetingSuccess());
-          fetchGreeting();
-          txnTimer.cancel();
-        }
-      });
+      if (type == LoginType.metaMask) {
```
Comment (kabaluyot) [inline lib/infrastructures/service/cubit/web3_cubit.dart]: use switch

File: lib/infrastructures/service/cubit/web3_cubit.dart [hunk 6]
```diff
@@ -71,36 +98,60 @@ class Web3Cubit extends Cubit<Web3State> {
...
+          text: text,
+          chainId: chainId,
+        );
+      }
     } catch (e) {
       emit(UpdateGreetingFailed(errorCode: '', message: e.toString()));
     }
   }
 
+  Future<void> sendTransaction({
```
Comment (kabaluyot) [inline lib/infrastructures/service/cubit/web3_cubit.dart]: remove this and add logic Transaction.callContract direct to 'updateGreeting'

File: lib/infrastructures/service/cubit/web3_cubit.dart (hunk 4 above)
Comment (kabaluyot) [inline lib/infrastructures/service/cubit/web3_cubit.dart]: Where is it invoked?

File: lib/infrastructures/service/cubit/web3_state.dart [hunk 7]
```diff
@@ -74,5 +74,27 @@ class UpdateGreetingFailed extends Web3State {
   final String message;
 }
 
+class InitializeGoogleProviderLoading extends Web3State {}
```
Comment (kabaluyot) [inline lib/infrastructures/service/cubit/web3_state.dart]: InitializeWeb3AuthProviderLoading ... etc

File: lib/module/auth/service/cubit/auth_state.dart [hunk 8]
```diff
@@ -53,3 +53,31 @@ class LoginWithMetamaskFailed extends AuthState {
...
+  const InitializeWeb3AuthFailed({
+    required this.errorCode,
+    required this.message,
+  });
+
+  final String errorCode;
+  final String message;
+}
+
+class LoginWithGoogleLoading extends AuthState {}
```
Comment (kabaluyot) [inline lib/module/auth/service/cubit/auth_state.dart]: LoginWithWeb3AuthGoogleSuccess

need separate vs tradiiton login with google

File: lib/internal/web3_utils.dart [hunk 9]
```diff
@@ -29,3 +40,51 @@ String getNetworkName(int chainId) {
       return 'Unknown Chain';
   }
 }
+
+enum LoginType {
+  metaMask,
+  google,
+}
+
+Uint8List bigIntToBytes(BigInt value, int length) {
```
Comment (kabaluyot) [inline lib/internal/web3_utils.dart]: move eip 191 and eip 712 signature related to

web3_eip191.dart
web3_eip171.dart

File: lib/internal/web3_utils.dart [hunk 10]
```diff
@@ -29,3 +40,51 @@ String getNetworkName(int chainId) {
...
+  Uint8List result = Uint8List(length);
```
Comment (kabaluyot) [inline lib/internal/web3_utils.dart]: Can we add eip191 and eip712 features in separate PR?

File: lib/infrastructures/service/cubit/web3_cubit.dart (hunk 3 above)
Comment (JosesGabriel) [reply lib/infrastructures/service/cubit/web3_cubit.dart]: Remove nalang nato ni pag mag implement na ta ug websocket? Dili sakto butangan ug pull to refresh if ani ang UI man gud. @kabaluyot 
image

File: lib/infrastructures/service/cubit/web3_cubit.dart (hunk 4 above)
Comment (JosesGabriel) [reply lib/infrastructures/service/cubit/web3_cubit.dart]: `home_screen.dart`

File: lib/infrastructures/service/cubit/web3_cubit.dart (hunk 3 above)
Comment (kabaluyot) [reply lib/infrastructures/service/cubit/web3_cubit.dart]: Shet karon ra nako nabasa my bad naka push nako sa removed periodic fetch greeting. Cge ra kols bahalag for internal use ra si pull to refresh for greeting para ma kita lang ang updated text without reopening ang app 

cc @JosesGabriel

**Review by kabaluyot — CHANGES_REQUESTED** [review]
- [x] pull to refresh for fetch greetings
- [x] fix unit tests
- [x] fix widget tests
- [x] QA in playstore internal build

**Review by kabaluyot — APPROVED** [review]

**Pull request merged**